include/turnstile.h

```diff
@@ -10,7 +10,8 @@ void turnstile_init(turnstile_t *);
 
 /* This puts currently running thread on the turnstile,
  * puts it in waiting state and yields. While waiting on turnstile
- * thread cannot be run. This function is done under critical section. */
+ * thread cannot be run. This function has to be called with critical section,
```

Comment:
Please don't change the semantics of this function. This is a bad practice that violates [Principle of Least Astonishment](http://www.catb.org/esr/writings/taoup/html/ch01s06.html#id2878339):

> The flip side of the Rule of Least Surprise is to avoid making things superficially similar but really a little bit different. This is extremely treacherous because the seeming familiarity raises false expectations. It's often better to make things distinctly different than to make them almost the same.

A user expects `turnstile_*` functions to be synchronized internally. But then... you make an exception for `turnstile_wait`, instead of just fixing the caller. If the user forgets to call `cs_enter` before, all hell will break loose without an apparent reason.

BTW: I hoped you'll ask about it instead of just reverting back to the original proposal.